[Sticky Header] Add 'Always' sticky option and a new setting to reduce logo sizes

LucasLacerdaUX opened the pull request:
**PR Summary:** 
- You can now customize header section to choose between `Never`, `On Scroll` and `Always` sticky modes.
- We've added a new setting to reduce logo size when the using sticky header. You can choose between `25%`, `50%` and `No reduction`.
- The maximum logo size has been updated from `250px` to `300px`

**Why are these changes introduced?**
Fixes #1454 and #1454

**What approach did you take?**
- `Reduce logo`: a `past-header` class is now added to header, so we can apply different styles to its elements. 
- `Always` Sticky header: 
  - Uses the same `sticky-header` component, with an early return for `hide` event.
  - I opted to re-use the component instead of simply using `position: sticky` without JS because we still need to detect when we have scrolled past the header.

**Demo links**
- [Store](https://os2-demo.myshopify.com/?preview_theme_id=128353304598)
- [Editor](https://os2-demo.myshopify.com/admin/themes/128353304598/editor)

**Checklist**
- [x] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Notified the [help.shopify.com](https://help.shopify.com) team about updates to theme settings (current contact: Kimli)
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

ludoboludo commented on assets/base.css:
```diff
@@ -2293,6 +2293,7 @@ input[type='checkbox'] {
 .header__heading-logo {
   height: auto;
   width: 100%;
+  transition: max-width var(--duration-short) ease;
```
this doesn't seem to work super well on my android phone. The animation is choppy: 

https://user-images.githubusercontent.com/28613840/173414220-e5850788-ed44-4d7d-951c-5d74ee8831eb.mp4

Oliviammarcello replied:
+1. Also the scaling down doesn't seem to be as accurate for mobile. 25% doesn't do anything and 50% feels quite minor. I know the 50% may be too dramatic of a jump, but I think it can reduce down a bit more. [Video](https://screenshot.click/14-13-tezfw-xorf3.mp4)

ludoboludo reviewed (commented):
It's working pretty well. I haven't found any bugs yet 👍 

One thing I'd like to bring up is the impact of these settings on the logo position. 

It's seem pretty useful and impactful on the `Middle left` position but less so on `Top left` and `Top center`. 
Feels like it's an ideal behaviour for the `Middle left` position but less so for the others 🤔 Though still useful I guess but seems less impactful or meant for those use cases imo.

LucasLacerdaUX commented on sections/header.liquid:
```diff
@@ -680,6 +697,10 @@
     }
 
     reset() {
+      if (this.getAttribute('data-sticky-type') === 'always') {
+        this.header.classList.remove('animate');
+        return;
+      }
```
Actually, the purpose of this line is to avoid removing the 'shopify-section-header-sticky' below.

Because it's always sticky, we shouldn't ever be removing this class so my idea was to do an early return here. 
But I just realized that we shouldn't be removing this `animate` class either? If it's always sticky :P

Probably a 
```js
      if (this.getAttribute('data-sticky-type') === 'always') return;
```
will be better.

Oliviammarcello commented:
- [ ] Katy suggested we change the "Reduce logo on scroll" help text to `Applies when sticky header is used.`

katycobb commented on sections/header.liquid:
```diff
@@ -830,17 +851,51 @@
+      ],
+      "default": "reduce-0",
+      "label": "t:sections.header.settings.reduce_logo_scroll.label",
+      "info": "t:sections.header.settings.reduce_logo_scroll.info"
```
Thanks for flagging this Ludo. I played around with it, and there are actually two settings here that only apply to logo images, so we should update both for clarity. I chatted with Olivia and we aligned on updating the setting labels:

- `Custom logo width` --> `Custom logo image width `
- `Reduce logo size on scroll` --> `Resize logo image on scroll` (Help text can remain as `Applies when sticky header is used.`) 
- Desktop logo position applies whether there is a logo image or not, so no change required here. 

Does that match your understanding?

ludoboludo replied:
💯 Perfect 👌 thanks for looking into it Katy 🙂

katycobb replied:
I tried to suggest the updates but I worried I did it wrong and deleted it 😨 Maybe I can ping you (or Lucas if he's feeling better) to pair for a few mins tomorrow. Would love to more confidently navigate these PRs and make the necessary updates rather than bugging you to do it!

katycobb commented on locales/en.default.schema.json:
```diff
@@ -918,6 +918,31 @@
+        },
+        "reduce_logo_scroll": {
+          "label": "Reduce logo size on scroll",
+          "info": "Only applies when sitkcy header is used.",
```
```suggestion
          "info": "Applies when sticky header is used.",
```

katycobb replied:
```suggestion
          "info": "Applies when sticky header is used.",
```

LucasLacerdaUX commented:
Thanks Katy, Ludo and Olivia for the reviews ❤️

I've just updated the label and help text for settings. Let me know if I missed anything. I'll look into the mobile resize issue this afternoon and request another review.

LucasLacerdaUX commented on sections/header.liquid:
```diff
@@ -830,17 +851,51 @@
+          "label": "t:sections.header.settings.sticky_header_type.options__3.label"
+        }
+      ],
+      "default": "none",
```
Good point. I think changing it to `scroll-up` would make sense, unless we wanted to change the default for some reason 👀

cc @Oliviammarcello

LucasLacerdaUX commented:
Sorry for taking so long, @ludoboludo and @Oliviammarcello.

200 years later, this is now ready for a second review. I've also requested translations now that the copy text has been updated. 

I've updated to use `width: 50%` and `width: 75%` instead of `max-width: calc({{ section.settings.logo_width }}px * 0.5);` and `max-width: calc({{ section.settings.logo_width }}px * 0.75);`.

This should fix the issue with the `Custom logo image width` setting causing the logo to not resize on mobile, due to actual width being smaller than the setting value.

I believe it also fixed the issue with choppy animation, but let me know if it didn't. It looks normal on my Android phone running the latest Chrome.

LucasLacerdaUX commented on assets/base.css:
```diff
@@ -2291,6 +2291,15 @@ input[type='checkbox'] {
+  transition: width var(--duration-short) ease;
+}
+
+.scrolled-past-header .header--reduce-25 .header__heading-logo {
+  width: 75%;
+}
+
+.scrolled-past-header .header--reduce-50 .header__heading-logo {
+  width: 50%;
```
I don't think we can use `transform` here because we need to reduce the actual area the logo will occupy, as the primary reason to reduce the logo on scroll is to make the sticky bar shorter.

`transform` simply resizes the image but takes the same space.

I can tweak the animation speed but I'm not sure we'll be able to avoid animating width 👀

ludoboludo replied:
Seems like on Express we did it by using the `max-width` and setting the `width` to either `100%` when the header is at the top and `auto` when sticky. 🤔 And Im not having an issue on my phone in terms of the animation.

LucasLacerdaUX commented on sections/header.liquid:
```diff
@@ -122,8 +122,8 @@
-<{% if section.settings.enable_sticky_header %}sticky-header{% else %}div{% endif %} class="header-wrapper color-{{ section.settings.color_scheme }} gradient{% if section.settings.show_line_separator %} header-wrapper--border-bottom{% endif %}">
-  <header class="header header--{{ section.settings.logo_position }} page-width{% if section.settings.menu != blank %} header--has-menu{% endif %}">
+<{% if section.settings.sticky_header_type != 'none' %}sticky-header data-sticky-type="{{ section.settings.sticky_header_type }}"{% else %}div{% endif %} class="header-wrapper color-{{ section.settings.color_scheme }} gradient{% if section.settings.show_line_separator %} header-wrapper--border-bottom{% endif %}">
+  <header class="header header--{{ section.settings.logo_position }} header--{{ section.settings.reduce_logo_scroll }} page-width{% if section.settings.menu != blank %} header--has-menu{% endif %}">
```
Good catch. Will add it 👍

Oliviammarcello reviewed (dismissed):
Looks good to me! 

On mobile after 160px the logo stops scaling up. @katycobb do you think we need help text to clarify that the "Custom logo image width" setting is optimised for mobile? [Video](https://screenshot.click/05-06-hhasw-nz06h.mp4)

katycobb commented:
> Looks good to me!
> 
> On mobile after 160px the logo stops scaling up. @katycobb do you think we need help text to clarify that the "Custom logo image width" setting is optimised for mobile? [Video](https://screenshot.click/05-06-hhasw-nz06h.mp4)

@Oliviammarcello
Hmm yep we should probably clarify this lil bit of magic. Something like this?

`Wider logos will be automatically optimized for mobile. `

Oliviammarcello commented:
> Reviewing it now in code, my feeling is that we allow a lot of flexibility for sizing, but are very opinionated on the resizing options (25-50-nothing) – would offering a stepper or range feel less opinionated and more flexible to the merchant? Not sure if there is an impact in the code or content,

@danielvan I do like this suggestion. Perhaps we could do [this](https://screenshot.click/13-25-li130-gdjt8.png)? It leaves more room for awkward combinations, but it does offer more flexibility and especially for the edge case of having a 300px width logo.

cc @LucasLacerdaUX

Oliviammarcello commented:
> * Olivia, I see that you have the default as `50%`. Shouldn't it be `0%` to keep the current behaviour (no resize on scroll) as default?

My bad, yes that's correct. 
 
> P.S.: I've temporarily updated to a different animation curve just to test if it would solve animation issues some folks were having on mobile devices, or if it's related to the `width` property itself.

I am no longer seeing the animation working for some reason 🤔

LucasLacerdaUX commented:
I've just updated the preview links, @Oliviammarcello.

Let me know if you still have issues with animations or the new setting.

https://os2-demo.myshopify.com/admin/themes/128353304598/editor?section=header
https://os2-demo.myshopify.com/?preview_theme_id=128353304598

danielvan commented:
I have tried something around these lines for the animation curve. Any thoughts?
My feeling is that it still feels like there's a delay before the resize kicks-off on scroll, but I'm unsure if there's a fix for it.  

The other element is I'm unsure if the nav items should realign with the logo, but maybe it could cause a lot of problems in navigations over multiple lines. 

[Video](https://screenshot.click/14-20-u8muz-1gvqh.mp4)

eugenekasimov commented:
I'm closing this PR ❗️

Here is a new PR #2165 as continue of the current one.
